### PR TITLE
✨ Implement e2e integration testing by using CAPI v1.6 test framework

### LINKED ISSUE #149: Add e2e validation for PRs
**User Story**

We should add e2e tests for PRs in addition to unit tests so we can verify if a change or version bump will break the overall project. This same job can also be run periodically as well.
 
[Miscellaneous information that will assist in solving the issue.]

/kind feature

### PR DESCRIPTION
**Pull request by Jont828**
**What this PR does / why we need it**: Finally got around to adding e2e testing. This will substantially improve the contributor process by providing a confident signal on whether or not a PR breaks anything in the e2e scenario.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #149

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -263,6 +264,16 @@ generate-modules: ## Run go mod tidy to ensure modules are up to date
 	go mod tidy
 	cd $(TOOLS_DIR); go mod tidy
 
+DOCKER_TEMPLATES := test/e2e/data/addons-helm
+
+.PHONY: generate-e2e-templates
+generate-e2e-templates: $(KUSTOMIZE)
```
Comment (mboersma) [inline Makefile:270]: Maybe add a `make help` string here?

File: Makefile [hunk 2]
```diff
@@ -263,6 +264,16 @@ generate-modules: ## Run go mod tidy to ensure modules are up to date
...
+	$(KUSTOMIZE) build $(DOCKER_TEMPLATES)/v1beta1/cluster-template --load-restrictor LoadRestrictionsNone > $(DOCKER_TEMPLATES)/v1beta1/cluster-template.yaml
+
+.PHONY: generate-flavors ## Generate template flavors.
```
Comment (mboersma) [inline Makefile:273]: I think `## Generate template flavors` should go on the next line (or does this work with `make help`)?

File: go.mod [hunk 3]
```diff
@@ -19,6 +19,7 @@ require (
 	k8s.io/klog/v2 v2.100.1
 	k8s.io/utils v0.0.0-20240102154912-e7106e64919e
 	sigs.k8s.io/cluster-api v1.6.0
```
Comment (mboersma) [inline go.mod:21]: Should this be `v1.6.1` in sync with `sigs.k8s.io/cluster-api/test`?

File: hack/common-vars.sh [hunk 4]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+# Copyright 2022 The Kubernetes Authors.
```
Comment (mboersma) [inline hack/common-vars.sh:2]:
```suggestion
# Copyright 2024 The Kubernetes Authors.
```

File: hack/gen-flavors.sh [hunk 5]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env bash
+
+# Copyright 2014 The Kubernetes Authors.
```
Comment (mboersma) [inline hack/gen-flavors.sh:3]:
```suggestion
# Copyright 2024 The Kubernetes Authors.
```

File: scripts/ci-e2e.sh [hunk 6]
```diff
@@ -14,87 +14,37 @@
...
+###############################################################################
 
 set -o errexit
+set -o nounset
 set -o pipefail
 
-REPO_ROOT=$(git rev-parse --show-toplevel)
-cd "${REPO_ROOT}" || exit 1
+REPO_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+# KIND="${REPO_ROOT}/hack/tools/bin/kind"
```
Comment (mboersma) [inline scripts/ci-e2e.sh:24]: We should remove these commented-out lines before merge.

File: test/e2e/common.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+const (
+	calicoHelmChartRepoURL   string = "https://docs.tigera.io/calico/charts"
+	calicoOperatorNamespace  string = "tigera-operator"
+	CalicoSystemNamespace    string = "calico-system"
+	CalicoAPIServerNamespace string = "calico-apiserver"
+	calicoHelmReleaseName    string = "projectcalico"
+	calicoHelmChartName      string = "tigera-operator"
+)
+
+// ensureControlPlaneInitialized waits for the cluster KubeadmControlPlane object to be initialized
```
Comment (mboersma) [inline test/e2e/common.go:83]:
```suggestion
// EnsureCalicoIsReady waits for the cluster KubeadmControlPlane object to be initialized
```

File: test/e2e/config/helm.yaml [hunk 8]
```diff
@@ -0,0 +1,144 @@
+managementClusterName: caaph-e2e
+
+images:
+  - name: ${MANAGER_IMAGE}
+    loadBehavior: mustLoad
+  - name: registry.k8s.io/cluster-api/cluster-api-controller:v1.6.0
```
Comment (mboersma) [inline test/e2e/config/helm.yaml:6]: Should we bump CAPI to v1.6.1 in general (or is that for a separate PR)?

File: test/e2e/e2e_suite_test.go [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+	Expect(dec.Decode(&e2eConfig)).To(Succeed())
+
+	// we unset Kubernetes version variables to make sure we use the ones resolved from the first Ginkgo ParallelNode in the e2e config.
+	os.Unsetenv(capi_e2e.KubernetesVersion)
+	os.Unsetenv(capi_e2e.KubernetesVersionUpgradeFrom)
+	os.Unsetenv(capi_e2e.KubernetesVersionUpgradeTo)
+
+	kubeconfigPath := parts[3]
+	bootstrapClusterProxy = framework.NewClusterProxy("bootstrap", kubeconfigPath, initScheme(), framework.WithMachineLogCollector(framework.DockerLogCollector{}))
+	// bootstrapClusterProxy = NewAzureClusterProxy("bootstrap", kubeconfigPath, framework.WithMachineLogCollector(AzureLogCollector{}))
```
Comment (mboersma) [inline test/e2e/e2e_suite_test.go:123]: Commented-out code should be removed before merge.

File: test/e2e/e2e_suite_test.go [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+	if !skipCleanup {
+		tearDown(bootstrapClusterProvider, bootstrapClusterProxy)
+	}
+})
+
+func loadE2EConfig(configPath string) *clusterctl.E2EConfig {
+	config := clusterctl.LoadE2EConfig(context.TODO(), clusterctl.LoadE2EConfigInput{ConfigPath: configPath})
+	Expect(config).NotTo(BeNil(), "Failed to load E2E config from %s", configPath)
+
+	// resolveKubernetesVersions(config)
```
Comment (mboersma) [inline test/e2e/e2e_suite_test.go:144]: Ditto here: more commented-out code.

File: test/e2e/e2e_suite_test.go [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+			Name:               config.ManagementClusterName,
+			RequiresDockerSock: config.HasDockerProvider(),
+			Images:             config.Images,
+		})
+		Expect(clusterProvider).NotTo(BeNil(), "Failed to create a bootstrap cluster")
+
+		kubeconfigPath = clusterProvider.GetKubeconfigPath()
+		Expect(kubeconfigPath).To(BeAnExistingFile(), "Failed to get the kubeconfig file for the bootstrap cluster")
+	} else {
+		// @sonasingh46: Workaround for testing workload identity.
```
Comment (mboersma) [inline test/e2e/e2e_suite_test.go:187]: We should probably remove the attribution to Ashutosh here.

File: test/e2e/config/helm.yaml (hunk 8 above)
Comment (Jont828) [reply test/e2e/config/helm.yaml:6]: Let's leave that for a separate PR (since we'd have to manually test all over again).

**Comment by Jont828** [comment]
@mboersma Just followed up on your comments. @nawazkh can you give this a review too before we get ready to merge?

File: test/e2e/config/helm.yaml [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+  default/wait-control-plane: ["10m", "10s"]
+  default/wait-worker-nodes: ["5m", "10s"]
+  default/wait-machine-pool-nodes: ["5m", "10s"]
+  default/wait-delete-cluster: ["3m", "10s"]
+  default/wait-machine-upgrade: ["20m", "10s"]
+  default/wait-machine-pool-upgrade: ["5m", "10s"]
+  default/wait-nodes-ready: ["10m", "10s"]
+  default/wait-machine-remediation: ["5m", "10s"]
+  default/wait-autoscaler: ["5m", "10s"]
+  default/wait-deployment: ["15m", "10s"]
```
Comment (nawazkh) [inline test/e2e/config/helm.yaml:128]: 🚀

File: test/e2e/data/addons-helm/v1beta1/bases/calico.yaml [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+        ipPools:{{range $i, $cidr := .Cluster.spec.clusterNetwork.pods.cidrBlocks }}
+        - cidr: {{ $cidr }}
+          encapsulation: VXLAN{{end}}
+      registry: mcr.microsoft.com/oss
+    # Image and registry configuration for the tigera/operator pod.
+    tigeraOperator:
+      image: tigera/operator
+      registry: mcr.microsoft.com/oss
+    calicoctl:
+      image: mcr.microsoft.com/oss/calico/ctl
```
Comment (nawazkh) [inline test/e2e/data/addons-helm/v1beta1/bases/calico.yaml:31]: nit: new line at the end of file

File: test/e2e/data/shared/v1beta1-provider/metadata.yaml [hunk 14]
```diff
@@ -0,0 +1,8 @@
+apiVersion: clusterctl.cluster.x-k8s.io/v1alpha3
+releaseSeries:
+  - major: 0
+    minor: 1
+    contract: v1beta1
+  - major: 0
+    minor: 2
+    contract: v1beta1
```
Comment (nawazkh) [inline test/e2e/data/shared/v1beta1-provider/metadata.yaml:8]: nit: needs new line at the end of the file

File: Makefile [hunk 15]
```diff
@@ -383,15 +394,34 @@ test-cover: ## Run unit and integration tests and generate a coverage report
...
+	$(GINKGO) -v --trace -poll-progress-after=$(GINKGO_POLL_PROGRESS_AFTER) \
+		-poll-progress-interval=$(GINKGO_POLL_PROGRESS_INTERVAL) --tags=e2e --focus="$(GINKGO_FOCUS)" \
+		$(_SKIP_ARGS) --nodes=$(GINKGO_NODES) --timeout=$(GINKGO_TIMEOUT) --no-color=$(GINKGO_NOCOLOR) \
+		--output-dir="$(ARTIFACTS)" --junit-report="junit.e2e_suite.1.xml" $(GINKGO_ARGS) $(ROOT_DIR)/$(TEST_DIR)/e2e -- \
+	    -e2e.artifacts-folder="$(ARTIFACTS)" \
+	    -e2e.config="$(E2E_CONF_FILE_ENVSUBST)" \
+	    -e2e.skip-resource-cleanup=$(SKIP_RESOURCE_CLEANUP) -e2e.use-existing-cluster=$(USE_EXISTING_CLUSTER)
+
+.PHONY: test-e2e-run
+test-e2e-run:
```
Comment (mboersma) [inline Makefile:416]: Could you add a `make help` comment here? And maybe clarify the one above for `test-e2e-run-skip-manifest`?

File: Makefile [hunk 16]
```diff
@@ -383,15 +394,34 @@ test-cover: ## Run unit and integration tests and generate a coverage report
...
+	$(MAKE) set-manifest-image MANIFEST_IMG=$(CONTROLLER_IMG)-$(ARCH) MANIFEST_TAG=$(TAG) TARGET_RESOURCE="./config/default/manager_image_patch.yaml"
+	$(MAKE) set-manifest-pull-policy TARGET_RESOURCE="./config/default/manager_pull_policy.yaml" PULL_POLICY=IfNotPresent
+	MANAGER_IMAGE=$(CONTROLLER_IMG)-$(ARCH):$(TAG) \
+	$(MAKE) test-e2e-run-skip-manifest
+
+.PHONY: get-e2e-kubeconfig
+get-e2e-kubeconfig:
```
Comment (mboersma) [inline Makefile:423]: Could you add a `make help` comment here?

File: scripts/ci-e2e.sh [hunk 17]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-# Copyright 2018 The Kubernetes Authors.
+# Copyright 2024 The Kubernetes Authors.
```
Comment (mboersma) [inline scripts/ci-e2e.sh:3]: I'm not a lawyer, but in general I don't think we should update the copyright assertion date in existing files, just use the current year when new files are added. (The dates don't have to agree between different files in the project.)

File: test/e2e/data/addons-helm/v1beta1/bases/calico.yaml [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+kind: HelmChartProxy
+metadata:
+  name: calico
+spec:
+  clusterSelector:
+    matchLabels:
+      cni: calico
+  repoURL: https://docs.tigera.io/calico/charts
+  chartName: tigera-operator
+  # version: ${CALICO_VERSION}
```
Comment (mboersma) [inline test/e2e/data/addons-helm/v1beta1/bases/calico.yaml:11]: Is this a comment or commented-out code?

File: test/e2e/e2e_suite_test.go [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+	return config
+}
+
+func createClusterctlLocalRepository(config *clusterctl.E2EConfig, repositoryFolder string) string {
+	createRepositoryInput := clusterctl.CreateRepositoryInput{
+		E2EConfig:        config,
+		RepositoryFolder: repositoryFolder,
+	}
+
+	// // Ensuring a CNI file is defined in the config and register a FileTransformation to inject the referenced file as in place of the CNI_RESOURCES envSubst variable.
```
Comment (mboersma) [inline test/e2e/e2e_suite_test.go:152]: Commented-out code here should be deleted IMHO.

File: test/e2e/e2e_suite_vars.go [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+	// Please note that provisioning will be skipped if e2e.use-existing-cluster is provided.
+	bootstrapClusterProvider bootstrap.ClusterProvider
+
+	// bootstrapClusterProxy allows to interact with the bootstrap cluster to be used for the e2e tests.
+	bootstrapClusterProxy framework.ClusterProxy
+
+	// kubetestConfigFilePath is the path to the kubetest configuration file
+	kubetestConfigFilePath string
+
+	// kubetestRepoListPath
```
Comment (mboersma) [inline test/e2e/e2e_suite_vars.go:71]: Could you flesh out the comment here?

**Comment by mboersma** [comment]
D'oh, I was too late. No worries, my comments were mostly cosmetic; I'll follow up with a PR to address them.

**Pull request merged**